In [1]:
#import libraries
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

In [2]:
uk = pd.read_csv('dataset\\uk_road_accident.csv')
uk

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

<h3>Descriptive Analytics</h3>

In [3]:
uk.describe()

Latitude      Longitude  Number_of_Casualties  Number_of_Vehicles
count  660654.000000  660653.000000         660679.000000       660679.000000
mean       52.553866      -1.431210              1.357040            1.831255
std         1.406922       1.383330              0.824847            0.715269
min        49.914430      -7.516225              1.000000            1.000000
25%        51.490690      -2.332291              1.000000            1.000000
50%        52.315641      -1.411667              1.000000            2.000000
75%        53.453452      -0.232869              1.000000            2.000000
max        60.757544       1.762010             68.000000           32.000000

In [4]:
uk.describe().T

count       mean       std        min        25%  \
Latitude              660654.0  52.553866  1.406922  49.914430  51.490690   
Longitude             660653.0  -1.431210  1.383330  -7.516225  -2.332291   
Number_of_Casualties  660679.0   1.357040  0.824847   1.000000   1.000000   
Number_of_Vehicles    660679.0   1.831255  0.715269   1.000000   1.000000   

                            50%        75%        max  
Latitude              52.315641  53.453452  60.757544  
Longitude             -1.411667  -0.232869   1.762010  
Number_of_Casualties   1.000000   1.000000  68.000000  
Number_of_Vehicles     2.000000   2.000000  32.000000

<h3>Checking for Null Values</h3>

In [5]:
uk.isnull()

Index  Accident_Severity  Accident Date  Latitude  Light_Conditions  \
0       False              False          False     False             False   
1       False              False          False     False             False   
2       False              False          False     False             False   
3       False              False          False     False             False   
4       False              False          False     False             False   
...       ...                ...            ...       ...               ...   
660674  False              False          False     False             False   
660675  False              False          False     False             False   
660676  False              False          False     False             False   
660677  False              False          False     False             False   
660678  False              False          False     False             False   

        District Area  Longitude  Number_of_Casualties  Number_of_Vehicles  \
0               False      False                 False               False   
1               False      False                 False               False   
2               False      False                 False               False   
3               False      False                 False               False   
4               False      False                 False               False   
...               ...        ...                   ...                 ...   
660674          False      False                 False               False   
660675          False      False                 False               False   
660676          False      False                 False               False   
660677          False      False                 False               False   
660678          False      False                 False               False   

        Road_Surface_Conditions  Road_Type  Urban_or_Rural_Area  \
0                         False      False                False   
1                         False      False                False   
2                         False       True                False   
3                         False      False                False   
4                         False       True                False   
...                         ...        ...                  ...   
660674                    False      False                False   
660675                    False      False                False   
660676                    False      False                False   
660677                    False      False                False   
660678                    False      False                False   

        Weather_Conditions  Vehicle_Type  
0                    False         False  
1                    False         False  
2                     True         False  
3                    False         False  
4                     True         False  
...                    ...           ...  
660674               False         False  
660675               False         False  
660676               False         False  
660677               False         False  
660678               False         False  

[660679 rows x 14 columns]

<h3>FILLING UP NULL VALUES</h3>

In [6]:
uk.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [7]:
#LATITUDE
uk['Latitude'].mean()
uk['Latitude'] = uk['Latitude'].fillna(uk['Latitude'].mean())

#LONGITUDE
uk['Longitude'].mean()
uk['Longitude'] = uk['Longitude'].fillna(uk['Longitude'].mean())
uk.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                       0
Light_Conditions               0
District Area                  0
Longitude                      0
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [8]:
#ROAD_SURFACE_CONDITION
uk['Road_Surface_Conditions'].mode()
uk['Road_Surface_Conditions'] = uk['Road_Surface_Conditions'].fillna(uk['Road_Surface_Conditions'].mode()[0])

#ROAD_TYPE
uk['Road_Type'].mode()
uk['Road_Type'] = uk['Road_Type'].fillna(uk['Road_Type'].mode()[0])

#URBAN_OR_RURAL_AREA
uk['Urban_or_Rural_Area'].mode()
uk['Urban_or_Rural_Area'] = uk['Urban_or_Rural_Area'].fillna(uk['Urban_or_Rural_Area'].mode()[0])
uk.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                       0
Light_Conditions               0
District Area                  0
Longitude                      0
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions        0
Road_Type                      0
Urban_or_Rural_Area            0
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [9]:
#WEATHER_CONDITION
uk['Weather_Conditions'].mode()
uk['Weather_Conditions'] = uk['Weather_Conditions'].fillna(uk['Weather_Conditions'].mode()[0])
uk.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

In [10]:
uk.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660679 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660679 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  660679 non-null  object 
 10  Road_Type                660679 non-null  object 
 11  Urban_or_Rural_Area      660679 non-null  object 
 12  Weather_Conditions       660679 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

<hr>

<h2>Questions:</h2>
<h3>1.What is the average number of casualties per accident?</h3>

In [18]:
uk['Number_of_Casualties'].mean()

np.float64(1.357040257068864)

<h3>Insight:</h3>
<h4>On average, each accident results in approximately 1.36 casualties, indicating a significant human impact per incident.</h4>
<hr>

<h3>2. How many accidents occurred on 'Dry' road surface conditions?</h3>

In [52]:
len(uk[uk['Road_Surface_Conditions'] == 'Dry'])

448547

<h3>Insight:</h3>
<h4>A large number of accidents, 448,547, occurred on dry road surfaces, suggesting that dry conditions do not necessarily guarantee safer driving.</h4>
<hr>

<h3>3. What is the most common 'Vehicle_Type' involved in accidents?</h3>

In [22]:
uk['Vehicle_Type'].value_counts().idxmax()

'Car'

<h3>Insight:</h3>
<h4>Cars are the most commonly involved vehicle type in accidents, highlighting their prevalence and potential risk on the roads.</h4>
<hr>

<h3>4. How many accidents occurred in 'Urban' areas?</h3>

In [23]:
uk[uk['Urban_or_Rural_Area'] == 'Urban'].shape[0]

421678

<h3>Insight:</h3>
<h4>With 421,678 accidents occurring in urban areas, it shows that city environments have a high frequency of traffic incidents.</h4>
<hr>

<h3>5. What is the average number of vehicles involved in accidents with 'Serious' severity?</h3>

In [24]:
uk[uk['Accident_Severity'] == 'Serious']['Number_of_Vehicles'].mean()

np.float64(1.6783273065282203)

<h3>Insight:</h3>
<h4>Accidents with serious severity typically involve about 1.68 vehicles on average, implying that multi-vehicle collisions are common in severe accidents.</h4>
<hr>

<h3>6. What is the average number of casualties in accidents with 'Fatal' severity?</h3>

In [25]:
uk[uk['Accident_Severity'] == 'Fatal']['Number_of_Casualties'].mean()

np.float64(1.9031289689412307)

<h3>Insight:</h3>
<h4>Fatal accidents have an average of about 1.9 casualties, reflecting the severe human cost of these incidents.</h4>
<hr>

<h3>7. Calculate the correlation coefficient between 'Number_of_Casualties' and 'Longitude'.</h3>

In [28]:
uk['Number_of_Casualties'].corr(uk['Longitude'])

np.float64(-0.040405645788454465)

<h3>Insight:</h3>
<h4>The very weak negative correlation (-0.04) between casualties and longitude suggests no meaningful geographic trend in casualties along the east-west axis.</h4>
<hr

<h3>8. Which are the top 5 'District Area' with the most accidents?</h3>

In [33]:
uk['District Area'].value_counts().head(5)

District Area
Birmingham    13491
Leeds          8898
Manchester     6720
Bradford       6212
Sheffield      5710
Name: count, dtype: int64

<h3>Insight:</h3>
<h4>Birmingham, Leeds, and Manchester are the top districts for accidents, indicating these urban centers have higher traffic incident rates.</h4>
<hr>

<h3>9. What is the percentage distribution of accidents across different 'Road_Surface_Conditions'?</h3>

In [39]:
(uk['Road_Surface_Conditions'].value_counts(normalize=True) * 100)

Road_Surface_Conditions
Dry                     67.891820
Wet or damp             28.260017
Frost or ice             2.802723
Snow                     0.891507
Flood over 3cm. deep     0.153933
Name: proportion, dtype: float64

<h3>Insight:</h3>
<h4>Most accidents (nearly 68%) occur on dry roads, showing that dry conditions are not necessarily safer for drivers.</h4>
<hr>

<h3>10. What is the percentage distribution of accidents across different 'Road_Type' categories?</h3>

In [40]:
(uk['Road_Type'].value_counts(normalize=True) * 100)

Road_Type
Single carriageway    75.174631
Dual carriageway      15.048760
Roundabout             6.658604
One way street         2.052283
Slip road              1.065722
Name: proportion, dtype: float64

<h3>Insight:</h3>
<h4>Single carriageway roads account for over 75% of accidents, highlighting their higher risk compared to other road types.</h4>
<hr>

<h3>11. What are the top 3 'Weather_Conditions' for accidents?</h3>

In [43]:
uk['Weather_Conditions'].value_counts().head(3)

Weather_Conditions
Fine no high winds       535013
Raining no high winds     79696
Other                     17150
Name: count, dtype: int64

<h3>Insight:</h3>
<h4>The majority of accidents happen in fine weather without high winds, suggesting that good weather does not eliminate accident risk.</h4>
<hr>

<h3>12. What is the percentage distribution of accidents in 'Urban' versus 'Rural' areas?</h3>

In [41]:
(uk['Urban_or_Rural_Area'].value_counts(normalize=True) * 100)

Urban_or_Rural_Area
Urban          63.824944
Rural          36.173391
Unallocated     0.001665
Name: proportion, dtype: float64

<h3>Insight:</h3>
<h4>About 64% of accidents occur in urban areas, emphasizing the higher traffic density and risk in cities.</h4>
<hr>

<h3>13. How many accidents had a 'Serious' severity?</h3>

In [54]:
len(uk[uk['Accident_Severity'] == 'Serious'])

88217

<h3>Insight:</h3>
<h4>There were 88,217 serious severity accidents, indicating a substantial number of high-impact incidents.</h4>
<hr>

<h3>14. How many accidents involved a 'Motorcycle over 500cc'?</h3>

In [55]:
uk[uk['Vehicle_Type'] == 'Motorcycle over 500cc'].shape[0]

25657

<h3>Insight:</h3>
<h4>Motorcycles over 500cc were involved in 25,657 accidents, showing their notable presence in traffic incidents.</h4>
<hr>

<h3>15. How does 'Number_of_Casualties' correlate with 'Latitude'?</h3>

In [56]:
correlation_casualties_latitude = uk['Number_of_Casualties'].corr(uk['Latitude'])
correlation_casualties_latitude 

np.float64(0.03220068662590628)

<h3>Insight:</h3>
<h4>The very weak positive correlation (0.03) between casualties and latitude indicates no significant north-south trend in casualties.</h4>
<hr>

<h3>16. How many accidents occurred under each 'Light_Conditions'?</h3>

In [63]:
light_conditions_counts = uk['Light_Conditions'].value_counts()
light_conditions_counts

Light_Conditions
Daylight                       484880
Darkness - lights lit          129335
Darkness - no lighting          37437
Darkness - lighting unknown      6484
Darkness - lights unlit          2543
Name: count, dtype: int64

<h3>Insight:</h3>
<h4>Most accidents occur during daylight (484,880), but a significant number also happen in darkness with lights lit, highlighting risks in various lighting conditions.</h4>
<hr>

<h3>17. Which 'District Area' has the most accidents?</h3>

In [64]:
district_accidents = uk['District Area'].value_counts().head(1)
district_accidents

District Area
Birmingham    13491
Name: count, dtype: int64

<h3>Insight:</h3>
<h4>Birmingham has the highest number of accidents among districts, pointing to it as a key area for road safety focus.</h4>
<hr>

<h3>18. How does 'Accident_Severity' vary across different 'Road_Type's?</h3>

In [65]:
severity_by_road_type = uk.groupby('Road_Type')['Accident_Severity'].value_counts(normalize=True).unstack()
severity_by_road_type

Accident_Severity      Fatal   Serious    Slight
Road_Type                                       
Dual carriageway    0.018255  0.118140  0.863604
One way street      0.007006  0.122059  0.870934
Roundabout          0.003228  0.083311  0.913462
Single carriageway  0.013208  0.142028  0.844764
Slip road           0.006959  0.086777  0.906263

<h3>Insight:</h3>
<h4>Single carriageways have the highest proportion of serious accidents (14.2%), suggesting these roads are more prone to severe crashes.</h4>
<hr>

<h3>19.  What is the most common 'Weather_Conditions' during accidents?</h3>

In [66]:
most_common_weather = uk['Weather_Conditions'].value_counts().head(1)
most_common_weather

Weather_Conditions
Fine no high winds    535013
Name: count, dtype: int64

<h3>Insight:</h3>
<h4>Fine weather with no high winds is the most common condition during accidents, indicating that accidents are not limited to adverse weather.</h4>
<hr>

<h3>20. How many accidents occurred when 'Weather_Conditions' were 'Raining no high winds'?</h3>

In [67]:
raining_accidents = uk[uk['Weather_Conditions'] == 'Raining no high winds'].shape[0]
raining_accidents

79696

<h3>Insight:</h3>
<h4>Nearly 79,700 accidents occurred during rain without high winds, showing that wet conditions still pose a significant hazard.</h4>
<hr>